### PR TITLE
Improve error handling

### LINKED ISSUE #95: Connection errors are silent
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure it has not already been reported

### Fastify version

3.18.0

### Plugin version

_No response_

### Node.js version

14.x

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

20.04

### Description

Fastify-redis rely on ioredis that doesn't exit the process when the connection can't be restored.

> emits when an error occurs while connecting.
However, ioredis emits all error events silently (only emits when there's at least one listener) so that your application won't crash if you're not listening to the error event.

We need to register an `error` handler on ioedis and graceful shutdown the application.

### Steps to Reproduce

Pass an invalid redis url or shutdown the redis instance during connection. No error is thrown but the app is still alive.

### Expected Behavior

If the application can't recover itself the instance must exit so that your orchestration software can't mark it as unhealthy and create a new one.

### PR DESCRIPTION
**Pull request by MSE99**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

#### Description

Closes #95 

This PR improves error handling, it will make the plugin wait until `Redis` instance fires `ready` event before calling the `next` function, it will also listen to the `error` event and propagate errors in case anything goes wrong on initial startup.

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't think this would work well as it would make call `next` twice/after boot is completed if the there is an error in the connection later on.

**Comment by mcollina** [comment]
I think you need to remove the `'error'` handler after `'ready'` is emitted.

**Comment by MSE99** [comment]
My bad, i thought it was possible to use the `next` function to propagate errors to fastify even after registering the plugin. removing the error event listener would mean that the user would have to manually close the client themselves after an error occurs.

**Review by mcollina — COMMENTED** [review]
Can you please not `return` the client?

File: index.js [hunk 1]
```diff
@@ -65,12 +65,21 @@ function fastifyRedis (fastify, options, next) {
   }
 
   if (!redisOptions.lazyConnect) {
-    return client.info((err, _) => {
-      return err ? next(err) : next()
-    })
-  } else {
-    next()
+    const onError = function (err) {
+      client.quit(() => next(err))
```
Comment (sf3ris) [inline index.js:69]: I don't think we should quit the client on the first error. In this case we are avoiding the retry strategy that is really important imho.
If we want to go with this `event` path I think we need to manually handle the retry strategy.

File: index.js [hunk 2]
```diff
@@ -65,12 +65,21 @@ function fastifyRedis (fastify, options, next) {
...
+      client.off('error', onError)
+      next()
+    }
+
+    return client
+      .on('ready', onReady)
+      .on('error', onError)
   }
+
+  next()
```
Comment (sf3ris) [inline index.js:82]: We want to call the `next()` without checking the connection only when the `lazyConnect` option has been provided right? 
In this case we are calling the `next()` here and another time when the client is in `ready`/`error` state.
Am i wrong?

**Review by sf3ris — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (climba03003) [reply index.js:82]: When specified `lazyConnect`, the if clause will stop at `return client` and this `next` is not reachable.

**Review by sf3ris — APPROVED** [review]
:+1: LGTM

File: index.js [hunk 3]
```diff
@@ -65,12 +65,44 @@ function fastifyRedis (fastify, options, next) {
...
+
+    const onError = function (err) {
+      // Swallow network errors to allow ioredis
+      // to preform reconnection and emit 'end'
+      // event if reconnection eventually
+      // fails.
+      // Any other errors during startup will
+      // trigger the 'end' event.
+      if (err instanceof Redis.ReplyError) {
+        this.emit('end', err)
```
Comment (mcollina) [inline index.js:93]: I would prefer if we did not call `emit()` on an object we do not maintain. Just call the `onEnd()` function.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**